scipy/stats/_wilcoxon.py

```diff
@@ -0,0 +1,261 @@
+import warnings
+import numpy as np
+
+from scipy import stats
+from ._stats_py import _normtest_finish
+from . import _morestats
+from ._axis_nan_policy import _broadcast_arrays
+from ._hypotests import _get_wilcoxon_distr
+from scipy._lib._util import _lazywhere
+
+
+class WilcoxonDistribution:
+
+    def __init__(self):
+        self._dists = {}
+
+    def _get_dist(self, n):
+        if n not in self._dists:
+            self._dists[n] = _get_wilcoxon_distr(n)
+        return self._dists[n]
+
+    def _cdf1(self, k, n):
+        pmfs = self._get_dist(n)
+        return pmfs[:k + 1].sum()
+
+    def _cdf(self, k, n):
+        return np.vectorize(self._cdf1, otypes=[float])(k, n)
+
+    def _sf1(self, k, n):
+        pmfs = self._get_dist(n)
+        return pmfs[k:].sum()
+
+    def _sf(self, k, n):
+        return np.vectorize(self._sf1, otypes=[float])(k, n)
+
+    def mean(self, n):
+        return n * (n + 1) // 4
+
+    def _prep(self, k, n):
+        k = np.asarray(k).astype(int, copy=False)
+        n = np.asarray(n).astype(int, copy=False)
+        mn = self.mean(n)
+        out = np.empty(k.shape, dtype=np.float64)
+        return k, n, mn, out
+
+    def cdf(self, k, n):
+        k, n, mn, out = self._prep(k, n)
+        return _lazywhere(k <= mn, (k, n), self._cdf,
+                          f2=lambda k, n: 1 - self._sf(k+1, n))[()]
+
+    def sf(self, k, n):
+        k, n, mn, out = self._prep(k, n)
+        return _lazywhere(k <= mn, (k, n), self._sf,
+                          f2=lambda k, n: 1 - self._cdf(k-1, n))[()]
+
+
+def rankdata(x, axis=-1):
```

Comment:
Should really rewrite `rankdata` and add `nan_policy`. Should document what `t` is and note that it left to correspond with ranks in ascending order, *not* reordered to correspend with original data / output `ranks`.